### ISSUE TITLE
Investigate using another CI system (Docker-based)

### ISSUE THREAD

**Issue by jakirkham** [jakirkham/nanshe#242]
It is time to investigate another CI system. Though Travis has been good to us in starting out, it is no longer feeling like the right fit. The main reason is builds have become quite slow ~10mins. Of that time nearly half (~4mins) is spent building the package, but most of it is redundant and remains the same over all versions. For instance, installing grid engine, installing miniconda, and activating a basic python conda environment. Even downloading and configuring some of our dependencies could probably be handled in advance.

One possible solution would be to consider binstar. However, it does not appear that they have considered security issues when interfacing with outside services like coveralls or push to GitHub.

The next solution would be a CI system that allows us to prepare our own containers for testing. The most popular format at the moment appears to be Docker Images. These can be built with a Dockerfile that easily fits in a typical VCS repo (e.g. git). They provide hooks for building an image when a push occurs. Also, they store the built images online at Docker Hub.

A few hosted CIs allow one to use any image from Docker Hub. The most notable include Wercker, Drone, and Circle CI. There may be others and certainly there are non-hosted ones. Due to the lightweight nature of containers and the popularity of Docker, it is reasonable to expect this list to grow. In fact, Travis does use Docker containers, but they don't allow us to customize them easily.

**Comment by jakirkham** [comment]
Some work on getting Docker containers up and going can be seen here ( https://github.com/jakirkham/docker_centos_drmaa_conda ) and here ( https://github.com/jakirkham/docker_ubuntu_drmaa_conda ). The CentOS version is further along as it has code that was borrowed from here ( https://bitbucket.org/deardooley/agave-docker ). This should give us the basics to install Grid Engine and DRMAA. Configuring them may still require some work, but will be aided by our Travis CI scripts included in this repo. Working with conda on a VM will be aided by the work here ( https://github.com/ContinuumIO/docker-images ).

**Comment by jakirkham** [comment]
Appears that Binstar also supports custom Docker containers ( https://github.com/Binstar/testci/blob/master/docker_test/README.md#test-out-the-new-binstar-build-with-docker ).

**Comment by jakirkham** [comment]
At this point, we have two fully function Docker containers (Ubuntu 12 and CentOS 6) flavors.

These have build scripts for reproducing them. They are automatically rebuilt and deployed to Docker Hub ( https://registry.hub.docker.com/ ) when they succeed.

Both have a fully functioning copy of Grid Engine installed with DRMAA support and configured to use out-of-the-box. Also, both have fully functioning installs of `conda` that include `conda-build` and `binstar` installed.

As a part of startup, they each have an entrypoint, which properly purge and then patch the system to handle the dynamic hostname, number of cpu cores, current user, etc. This ensures there are no remnants from tests performed during building and that everything is ready to go when launched. Both provide a default `bash` command to allow one to easily enter the box and begin working.

**Comment by jakirkham** [comment]
Working implementation now can be used on Wercker. It is simply a matter of setting up environment variables for deployment and such.

**Comment by jakirkham** [comment]
Tried Drone, Shippable, and Wercker. Drone's web version is insufficient for our needs and does not yet support the in repo YAML file. Shippable is simple coming from Travis. However, they only allow one build at a time and we ran into problems with out containers. Wercker has good Docker support, allows offline building for free, also any build can be pulled as a new Docker image for inspection (useful when it fails). Hence, we will use Wercker at present.
